Use `{processx::run}` to capture `cargo` output

Ilia-Kosenkov opened the pull request:
This is an attempt to fix #115. Currently, we get little to no information about the reason the compilation has failed.
Now we can throw an error that contains the output of `cargo` marking problematic code fragments.
``` r
rextendr::rust_function("fn wrong_function(x : i33) -> f33 { false }")
#> i build directory: 'C:/Users/[redacted]/AppData/Local/Temp/RtmpM11rtn/file2e184e2156d4'
#> Error: Rust code could not be compiled successfully. Aborting.
#> error[E0412]: cannot find type `i33` in this scope
#>  --> src\lib.rs:3:23
#>   |
#> 3 | fn wrong_function(x : i33) -> f33 { false }
#>   |                       ^^^ help: a builtin type with a similar name exists: `i32`
#> 
#> 
#> error[E0412]: cannot find type `f33` in this scope
#>  --> src\lib.rs:3:31
#>   |
#> 3 | fn wrong_function(x : i33) -> f33 { false }
#>   |                               ^^^ help: a builtin type with a similar name exists: `f32`
#> 
#> 
#> error: aborting due to 2 previous errors
```

<sup>Created on 2021-12-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>


![Screenshot 2021-12-12 232131](https://user-images.githubusercontent.com/8782986/145728271-34578e22-220c-4ec7-8336-045f20b0c23c.png)

malcolmbarrett commented:
Excited about this one!

Ilia-Kosenkov commented:
@yutannihilation, Windows builds seem to fail because of the missing `-lbcrypt`. Can you make a PR to fix this so I can then merge master and re-run CI?
https://github.com/extendr/extendr/issues/339

Ilia-Kosenkov commented on R/source.R:
```diff
@@ -282,22 +270,63 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
+
+  if (!isTRUE(compilation_result$status == 0)) {
+    errors <-
+      message_buffer %>%
+      purrr::map_chr(stringi::stri_trim) %>%
+      purrr::map(jsonlite::parse_json) %>%
+      purrr::keep(
+        ~.x$reason == "compiler-message" && .x$message$level == "error"
+      )
```
Here I rely on the fact that `cargo` output is configured to produce JSONs. Should we have an extra `tryCatch` to handle a situation when the output is garbage and can't be parsed as JSON?

Ilia-Kosenkov commented:
Also, the devel version of R seems to produce weird knitr output, will research it later.
https://github.com/extendr/rextendr/runs/4499559708?check_suite_focus=true#step:11:182

yutannihilation commented:
Here's the PR: https://github.com/extendr/extendr/issues/339

Ilia-Kosenkov commented:
So I realized that if an error message is too long (by default it is `options(warning.length = 1000)`, and measured in regular characters, including various invisible ANSI style delimiters), the text gets truncated, and if truncation corrupts ANSI sequence, the terminal gets screwed up (similar to what is shown in the image at the beginning of this thread). The limit can be increased up to I think `8170`, which will be automatically done by `rlang::abort` in the future https://github.com/r-lib/rlang/pull/1214. Still, 8000 is not that much, so first of all I implemented an algorithm that cuts off compiler messages if they no longer fit in the output buffer. The attached image shows how it looks like with a limit of `1000` characters.

![image](https://user-images.githubusercontent.com/8782986/146444247-5656aeae-5fff-4f45-9d68-fe523d6bb00d.png)

The second step is to locally override options to set `warning.length` to `8000` so we can display large error messages. This should be enough for the majority of cases, when code fragment has one to five (I guess) reported errors. Same fragment but with a limit of `8000`:

![image](https://user-images.githubusercontent.com/8782986/146444369-2b931e4c-7b04-40ed-8d22-0da7effe5515.png)

The first 'could not compile' error message actually comes from the output of compilation process, and can be silenced with `quiet = TRUE`. Everything else is an error that can be re-inspected with `rlang::last_error()` or `rlang::last_trace()`:

![image](https://user-images.githubusercontent.com/8782986/146444646-1420a664-6cc8-46c6-b3af-9297effb25d0.png)

Ilia-Kosenkov commented:
Added `x` to the beginning of each 'error' in the 'details' list.
![image](https://user-images.githubusercontent.com/8782986/146445597-ea31b694-c6bf-4cd8-b61d-1dc99d3989d1.png)

Ilia-Kosenkov commented:
The changes are significant, I'd like an honest review from as many contributors as possible.

clauswilke commented on tests/testthat/_snaps/unix/helper-functions.md:
```diff
@@ -0,0 +1,25 @@
+
+    Rust code could not be compiled successfully. Aborting.
+    x error[E0412]: cannot find type `i33` in this scope
+     --> src/lib.rs:3:19
```
Are you sure you want to snapshot these error messages? Any time something minor changes in rustc's output this will break.

Ilia-Kosenkov replied:
I understand the danger, yet I trust Rust developer team and believe they will not change the output of `cargo`/`rustc`. Even if they will (e.g., if something like Rust 2.0 is released), I see no problem in updating snapshots. After all, every time `{rlang}`, `{cli}`, `{knitr}`, or any other library that produces formatted output has something changed under the hood, we are forced to update the snapshots (as soon as we notice). This also draws our attention to potential changes that we overlook.

clauswilke replied:
Ok, sounds good.

clauswilke commented on R/source.R:
```diff
@@ -282,22 +270,65 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
+      sprintf("--target-dir=%s", file.path(dir, "target")),
+      if (profile == "release") "--release" else NULL,
+      "--message-format=json-diagnostic-rendered-ansi",
+      ifelse(
```
Why `ifelse` instead of `if () ... else`? `ifelse` is normally used for vectorized data. Also, it's inconsistent with line 292.

Ilia-Kosenkov replied:
Fixed this and a few other arguments. I think `{glue}` works better here than `sprintf()`.

clauswilke reviewed (commented):
Overall looks good to me. A few minor comments.

Ilia-Kosenkov commented on R/source.R:
```diff
@@ -118,16 +118,6 @@ rust_source <- function(file, code = NULL,
     ui_i("build directory: {.file {dir}}")
-
-    # `""` sends `cargo` output to R's standard output.
-    # R console displays to the user informaion about compilation steps and
```
I think it was present in the original code fragment, but I deleted it.

clauswilke commented:
Any idea what happened with the last series of commits, where now we don't see line-by-line changes anymore for file `R/source.R`?

Ilia-Kosenkov commented:
Ok, I screwed up line endings, will fix it tomorrow.

Ilia-Kosenkov commented:
Also, I think we are now missing compilation warnings (like when you have unused variables or parameters). It is not critical, but I'd like to capture that information as well and display it to the user. It should not be difficult, as we already have all the output in JSON. We just need to filter warnings and print them using `{cli}`. I'd like to implement it in this PR.

Ilia-Kosenkov commented:
So I fixed the issue with line endings, started displaying warnings emitted by cargo, and also fixed a small bug that displayed one line of output even if `quiet = TRUE` was set. Now any interactive compilation is completely silent when `quiet = TRUE` is provided.

Ilia-Kosenkov commented:
![image](https://user-images.githubusercontent.com/8782986/146676545-60d1d3b3-8edd-4e1b-8f6e-55236e39633b.png)

yutannihilation commented:
Looks awesome! I'll try to find time to review this.

Just for information, processx is currently broken on UCRT R-devel on Windows. I'm suspecting the problem is something related to the difference between the OS's encoding (not necessarily UTF-8) and the R's native encoding (UTF-8), so rextendr might be affected. It is probably a bit too early to worry about this (and we don't have CI runner for this), but we might need to have better understanding about the encoding and the stdout/stderr of an external process.

https://cran.r-project.org/web/checks/check_results_processx.html

Ilia-Kosenkov commented:
When we generate bindings out of process, we use `{callr}`, which AFAIK relies on `{processx}`, so we are already affected. https://github.com/r-lib/callr/blob/ea5c3df7788e43870a265b276d3a4ae103e91557/DESCRIPTION#L20

yutannihilation commented:
Oh, I see. I think we should add an Windows runner with R-devel to our CI. It seems the UCRT is getting ready.

c.f. https://developer.r-project.org/Blog/public/2021/12/07/upcoming-changes-in-r-4.2-on-windows/index.html

I'll take a look on this.

Ilia-Kosenkov commented:
Also it would be nice if you can migrate the second CI pipeline to r-actions instead of us doing all the heavy lifting, similar to R CMD check pipeline

yutannihilation commented on R/source.R:
```diff
@@ -282,22 +269,117 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
-    stderr = stderr
+    echo_cmd = FALSE,
+    windows_verbatim_args = FALSE,
+    stderr = ifelse(isTRUE(quiet), "|", ""),
```
You can also replace `ifelse()` with `if` here.

yutannihilation commented on R/source.R:
```diff
@@ -282,22 +269,117 @@ invoke_cargo <- function(toolchain, specific_target, dir, profile,
+#' @noRd
+check_cargo_output <- function(compilation_result, message_buffer, tty_has_colors, quiet) {
+  cargo_output <- message_buffer %>%
+    purrr::map_chr(stringi::stri_trim) %>%
```
Maybe we can parse JSONs without trimming whitespaces? Just curious.

Ilia-Kosenkov replied:
You were right, it seems to work fine.

yutannihilation commented on R/ui.R:
```diff
@@ -103,17 +103,66 @@ ui_w <- function(text = "", env = parent.frame()) {
+subset_lines_to_fit_in_limit <- function(lines,
+                                         max_length = 1000,
+                                         truncation_notification = "{.val {n_removed}} compiler messages not shown.") {
+  n_removed <- 100L
```
```suggestion
```

It looks this line is unnecessary.

Ilia-Kosenkov replied:
This may be an overkill, but the trick here is to first compute the actual size of the message appended if lines are trimmed. This message is i) customizable ii) depends on the number of removed lines (which is unknown).
So I first assign an arbitrarily large value to `n_removed`, then I use `{cli}` to obtain the actual interpolated string and then use it to compute the length. Assuming we will never have more than 999 lines removed, it gives an upper limit on the length of the added string, which is enough to fit everything within the buffer.
If I remove this declaration, the `{cli}` will fail interpolation.

It is sort of a circular dependency -- `n_removed` depends on the length of the appended message, which in turn depends on `n_removed`.

yutannihilation replied:
Ah, sorry, I just missed this variable is referenced by `truncation_notification`. Thanks for the explanation.

clauswilke replied:
I would suggest to add a few comments to the code explaining this reasoning. Whenever we have to do some weird trick to get something done, it's a good idea to write the reasoning down, as nobody will remember five years from now.

yutannihilation commented on R/ui.R:
```diff
@@ -103,17 +103,66 @@ ui_w <- function(text = "", env = parent.frame()) {
+                                         max_length = 1000,
+                                         truncation_notification = "{.val {n_removed}} compiler messages not shown.") {
+  n_removed <- 100L
+  truncation_notification_size <- nchar(bullet_i(truncation_notification))
```
By default, `nchar()` measures the numbers of the characters, but it seems it's actually measured in bytes (see the last one is trancated by 334 repeats).

```
> cat("\u2705")
✅> 
> nchar("\u2705")
[1] 1
> nchar("\u2705", type = "width")
[1] 2
> nchar("\u2705", type = "byte")
[1] 3
> 
> withr::with_options(
+   list(warning.length = 1000),
+   warning(stringr::str_dup("\u2705", 333))
+ )
Warning message:
In force(code) :
  ✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅
> 
> withr::with_options(
+   list(warning.length = 1000),
+   warning(stringr::str_dup("\u2705", 334))
+ )
Warning message:
In force(code) :
  ✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅✅ [... truncated]
```

Ilia-Kosenkov replied:
Interesting, the documentation on this topic is not as good as I expected, so I assumed we dealt with characters. I will use `type = "byte"`.

yutannihilation reviewed (commented):
Generally looks good to me. Minor comments.

Ilia-Kosenkov commented:
I am getting a weird error https://github.com/extendr/rextendr/runs/4592571595?check_suite_focus=true#step:8:368 on Windows, and it does not seem to be reproduced locally. I wonder if it happens here 
```r
 rtools_home <- normalizePath(
      file.path(pkgbuild::rtools_path(), "..", ".."),
      winslash = "/",
      mustWork = TRUE
    )```
and if `pkgbuild::rtools_path()` detects two rtools, `v40` and `v42` for ucrt? Is it possible, @yutannihilation?

Ilia-Kosenkov commented:
Oh yeah, that's `{pkgbuild}` fault https://github.com/r-lib/pkgbuild/releases/tag/v1.3.1

Ilia-Kosenkov commented:
https://github.com/r-lib/pkgbuild/commit/2a9ce8d9c25c6dafb29ecb528247f49c10d8c59a#diff-5d9310c23b0c279cdd57b95238eb945cd45bac174c076244ad73cd4147deb626
https://github.com/r-lib/pkgbuild/blob/2a9ce8d9c25c6dafb29ecb528247f49c10d8c59a/R/rtools-metadata.R#L66-L69

yutannihilation commented:
Oh, I forgot there's `ucrt64\bin`  bundled in Rtools. Your fix looks good to me.

Btw, we probably need to tweak this to handle the case of `RTOOLS42_HOME`.

https://github.com/extendr/rextendr/blob/9b2f49519b7b86a5b31f8ad20cd652566d43c498/R/source.R#L283

Ilia-Kosenkov commented:
I think this will be needed for R 4.2.0+. I was hoping you can review and improve this when you investigate running extendr targeting `ucrt` :) I can help you with testing though

yutannihilation commented:
Yeah, I'll take a look. I'm now using R-devel as my default environment so I'm almost ready for testing it! (except for uninstalling rtools40).

malcolmbarrett commented:
I actually can't test this currently, which I think is directly related to #100. I've had trouble with my `PATH` and cargo with processx in the path 🤔

malcolmbarrett commented:
I wonder if the `env` argument of `processx::run()` can somehow include the local `PATH` if it set up as in #99

Ilia-Kosenkov commented:
I think I have addressed all the comments. Should I merge?

clauswilke commented:
No concerns from my side.

malcolmbarrett commented:
Once you merge this, people in my boat, e.g. who have #99 setup on Macs, may not be able to use the package (as I cannot with this PR). I don't necessarily think it's the responsibility of this PR to address it, but once merged, that issue and #100 become critical.

clauswilke commented:
Actually, sorry for not connecting the dots earlier. I think @malcolmbarrett is making a good point. If this PR completely bricks the software for people with a reasonable setup (having a path set in `.Renviron` is totally reasonable) we have a problem.

Ilia-Kosenkov commented:
Ok, I see. I suggest to switch this issue to draft and then investigate processx misbehavior. I have no Apple devices in my household, so it is difficult to debug this issue. I will try to find a MacOS VM to reproduce the problem and find a solution.

yutannihilation commented:
I'd suggest merging this. As this is the development version, I think it's okay if the `HEAD` gets temporarily broken. As I commented on https://github.com/extendr/rextendr/issues/100#issuecomment-1000564058, I think the only choice is to add `PATH` wherever we use `cargo` (we need to fix the `Makevars` template as well). It probably takes some more changes, not all of which is related to this issue, so I want to separate the discussion, if this sounds okay to @malcolmbarrett.

malcolmbarrett commented:
Yes, I think that's reasonable. This PR definitely escalates the issue, but I don't think it should necessarily be responsible for addressing it. Your solution looks compelling, @yutannihilation!

yutannihilation commented:
Thanks for confirming!

Ilia-Kosenkov commented:
Then I will merge it later today, thanks for the discussion.
